### ISSUE TITLE
Generate the state with data on the request

### ISSUE THREAD

**Issue by rmiraballes** [fastify/fastify-oauth2#40]
## 🚀 Feature Proposal

Generate the state with information on the request.

## Motivation

I have information on the queryParams that I need it when the request returns from the OAuth flow. The `generateStateFunction` does not receive any parameters and in` callbackUriParams` there is not access to the request.

## Example
```javascript
 fastify.register(oauthPlugin, {
 ...
 generateStateFunction: (request) => {
      // Do something with the request and return a state
      return request.query.code
  }
 })
```
I can create the pr for this if this is aligned with the idea of the library. Basically is to inject the request to the  `generateStateFunction` here: [code]( https://github.com/fastify/fastify-oauth2/blob/master/index.js#L58
)

**Comment by Eomm** [comment]
Just to know: is `request.query.code' for the openId purpose?

I think it is fine adding a parameter to that function, please don't forget the tests

**Comment by rmiraballes** [comment]
Thanks for the quick response! 

No is not, I just need to create the state from some information that I have on the request, that was just an example.

I'll create the PR.

**Comment by Eomm** [comment]
Released in 3.2.0

Thks!
